neqo-transport/src/events.rs

```diff
@@ -99,23 +111,131 @@ impl ConnectionEvents {
+        // If 0rtt rejected, must start over and existing events are no longer
+        // relevant.
         self.events.borrow_mut().clear();
         self.insert(ConnectionEvent::ZeroRttRejected);
     }
 
     pub fn events(&self) -> impl Iterator<Item = ConnectionEvent> {
-        self.events.replace(BTreeSet::new()).into_iter()
+        self.events.replace(VecDeque::new()).into_iter()
+    }
+
+    pub fn has_events(&self) -> bool {
+        !self.events.borrow().is_empty()
     }
 
+    pub fn next_event(&self) -> Option<ConnectionEvent> {
+        self.events.borrow_mut().pop_front()
+    }
+
+    #[allow(clippy::block_in_if_condition_stmt)]
     fn insert(&self, event: ConnectionEvent) {
-        self.events.borrow_mut().insert(event);
+        let mut q = self.events.borrow_mut();
+
+        // Special-case two enums that are not strictly PartialEq equal but that
+        // we wish to avoid inserting duplicates.
+        if match &event {
+            ConnectionEvent::SendStreamStopSending { stream_id, .. } => q.iter().any(|evt| {
+                matches!(
+		    evt, ConnectionEvent::SendStreamStopSending { stream_id: x, .. }
+		    if *x == *stream_id)
+            }),
+            ConnectionEvent::RecvStreamReset { stream_id, .. } => q.iter().any(|evt| {
+                matches!(
+		    evt, ConnectionEvent::RecvStreamReset { stream_id: x, .. }
+		    if *x == *stream_id)
+            }),
+            _ => q.contains(&event),
+        } {
+            // Already in event list.
+            return;
+        }
+
+        q.push_back(event);
     }
 
-    pub fn has_events(&self) -> bool {
-        !self.events.borrow().is_empty()
+    fn remove<F>(&self, f: F)
+    where
+        F: Fn(&ConnectionEvent) -> bool,
+    {
+        self.events.borrow_mut().retain(|evt| !f(evt))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::{ConnectionError, Error};
+
+    #[test]
+    fn event_culling() {
+        let evts = ConnectionEvents::default();
+
+        evts.client_0rtt_rejected();
+        evts.client_0rtt_rejected();
+        assert_eq!(evts.events().count(), 1);
+        assert_eq!(evts.events().count(), 0);
+
+        evts.new_stream(4.into());
+        evts.new_stream(4.into());
+        assert_eq!(evts.events().count(), 1);
+
+        evts.recv_stream_readable(6.into());
+        evts.recv_stream_reset(6.into(), 66);
+        evts.recv_stream_reset(6.into(), 65);
+        assert_eq!(evts.events().count(), 1);
+
+        evts.send_stream_writable(8.into());
+        evts.send_stream_writable(8.into());
+        evts.send_stream_stop_sending(8.into(), 55);
+        evts.send_stream_stop_sending(8.into(), 56);
+        let events = evts.events().collect::<Vec<_>>();
+        assert_eq!(events.len(), 1);
+        assert_eq!(
+            events[0],
+            ConnectionEvent::SendStreamStopSending {
+                stream_id: 8,
+                app_error: 55
+            }
+        );
```

Comment:
nit: you could also check that the event that remains in the queue is on its original place: if we have event 'b' that is not related to stream 8 and it is inserted between two send_stream_stop_sending that the order will be:
send_stream_stop_sending{stream_id: 8, ..},
send_stream_writable{ stream_id: x,}